apis/v1alpha2/httproute_types.go

```diff
@@ -320,6 +321,13 @@ const (
 	HeaderMatchImplementationSpecific HeaderMatchType = "ImplementationSpecific"
 )
 
+// HTTPHeaderName is the name of an HTTP header.
+//
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=256
+// +kubebuilder:validation:Pattern=`^[-A-Za-z0-9]+$`
```

Comment:
this forbids pseudo headers and a bunch of other valid things like periods, slashes, percent signs, etc. rfc7230 3.2.6 describes the full list

Reply:
Interesting, are you able to understand from the RFC which characters are valid for a header name? It's hard for me to distinguish, since that specific section seems to be focused on the field value? If the `token` represents the name, it seems like the allowed set of chars would be a-z0-9 + this:

```
"!" / "#" / "$" / "%" / "&" / "'" / "*" / "+" / "-" /  "." / "^" / "_" / "`" / "|" / "~" 
```

Then to further complicate things, the spec seems to explicitly rule out `:` as a value, which seems like it would be an issue for pseudo headers...

Reply:
For reference, here's where the current regex is coming from: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/validation/validation.go#L388-L400. You're completely right that it would prevent pseudo headers though.

Reply:
I think all of those are valid.

The trick for pseudo header is that `:` is also the separator for the k/v - so in some cases it ends up being treated as `""` as the key. Not sure if its standard though: 
```
$ curl httpbin.org/get -H ':foo: bar'
{
  "args": {},
  "headers": {
    "": "foo: bar"
```

Reply:
Also the RFC is hard to follow but if you look at https://datatracker.ietf.org/doc/html/rfc7230#section-3.2, a header key is a `token`, and a token is one of those you listed above. Looking at the golang source was easier than the RFC...

Reply:
So to clarify, you're saying that we wouldn't need `:` to be a valid header name character?

Reply:
If we want to support only _real_ headers - no `:`.

If we want to support pseudo headers though, then we do want it. So that would allow header match on `:path`, `:scheme`, etc.

I think either one is valid, but we should be deliberate. If all pseudoheaders are representable based on other first class fields, then its plausble we do not allow them

Reply:
I think there's been at least some discussion around pseudo header matching being useful in some cases, but since I don't think we've reached any consensus on that, it seems safer to leave it out with the possibility of adding it later if necessary.

Reply:
Allowing pseudoheader pros:
- allows people to quickly try out some more advanced routing things, before the API supports it.
Cons:
- less incentive to have the API support it, because you can use pseudoheaders.

I think it would be better long-term to have the functionality explicitly defined rather than do the pseudoheader thing, but it's not a strong preference.

Reply:
Extended regex to include characters listed above, avoided pseudo headers for now as support will be much easier to add than remove in the future.